n_cjson_helpers.c

```diff
@@ -431,15 +431,18 @@ const char *JGetItemName(const J * item)
 void JItoA(long int n, char *s)
 {
     char c;
-    long int i, j, sign;
-    if ((sign = n) < 0) {
-        n = -n;
+    // Conversion to unsigned is required to handle the case where n is
+    // LONG_MIN.
```

Comment:
This comment is not obvious. Why is this the case?

Reply:
If n is `JINTEGER_MIN`, when we negate it with the unary minus operator, it overflows and the result is undefined behavior (I'm pretty sure). If I keep the original code that was here and pass in `LONG_MIN`, this

```
n = -n;
```

Just keeps n as `LONG_MIN`. Then you end up doing `n % 10` on a negative number and the whole algorithm breaks down. If we cast n to unsigned, then the rules for negation are different. See: https://stackoverflow.com/questions/8026694/c-unary-minus-operator-behavior-with-unsigned-operands

> The negative of an unsigned quantity is computed by subtracting its value from 2^n, where n is the number of bits in the promoted operand.

So `unsignedN = -unsignedN` when `JINTEGER` is 32 bits and n is `JINTEGER_MIN` yields 2^32-2147483648, which is 2147483648. Then the '% 10' algorithm works as intended.

I'll explain all this better in an improved code comment.

Reply:
That explanation makes more sense when the data type is signed. I don't see how this works with an unsigned data type.

Reply:
`JINTEGER_MIN` should be `0x80000000`, right?

If you make that unsigned, then you have a valid positive number.
`unsigned long int unsignedN = n;`

Then if you make it negative with `-`. Now, you have a temporary 64-bit negative number.
`0xFFFFFFFF80000000`, which you then assign to a 32-bit unsigned number, so you are right back to a valid 32-bit unsigned number `0x80000000`.

Reply:
I understand that modulo will work better on unsigned values, but I don't quite get the point of setting `unsignedN` to `-unsignedN`.